contracts/test/PlonkVerifier.t.sol

```diff
@@ -0,0 +1,432 @@
+    /// @dev Test constants declared matches that from Jellyfish
+    function test_correctConstants() external {
+        string[] memory cmds = new string[](3);
+        cmds[0] = "diff-test";
+        cmds[1] = "plonk-constants";
+
+        bytes memory result = vm.ffi(cmds);
+        (
+            uint256 k1,
+            uint256 k2,
+            uint256 k3,
+            uint256 k4,
+            uint256 betaHX0,
+            uint256 betaHX1,
+            uint256 betaHY0,
+            uint256 betaHY1
+        ) = abi.decode(
+            result, (uint256, uint256, uint256, uint256, uint256, uint256, uint256, uint256)
+        );
+
+        assertEq(k1, V.COSET_K1);
+        assertEq(k2, V.COSET_K2);
+        assertEq(k3, V.COSET_K3);
+        assertEq(k4, V.COSET_K4);
+        assertEq(betaHX0, V.BETA_H_X0);
+        assertEq(betaHX1, V.BETA_H_X1);
+        assertEq(betaHY0, V.BETA_H_Y0);
+        assertEq(betaHY1, V.BETA_H_Y1);
+    }
+}
+
+contract PlonkVerifier_batchVerify_Test is PlonkVerifierCommonTest {
+    /// forge-config: default.fuzz.runs = 30
+    /// @dev Test if some of the user inputs are invalid
+    function testFuzz_revertWhenInvalidArgs(
+        IPlonkVerifier.VerifyingKey[] memory verifyingKeys,
+        uint256[][] memory publicInputs,
+        IPlonkVerifier.PlonkProof[] memory proofs,
+        bytes[] memory extraTranscriptInitMsgs
+    ) external {
+        vm.assume(
+            verifyingKeys.length != proofs.length || publicInputs.length != proofs.length
+                || extraTranscriptInitMsgs.length != proofs.length || proofs.length == 0
+        );
+        vm.expectRevert(V.InvalidPlonkArgs.selector);
+        V.batchVerify(verifyingKeys, publicInputs, proofs, extraTranscriptInitMsgs);
+    }
+
+    /// @dev Test happy and unhappy path of `batchVerify`.
+    function test_batchVerify_succeeds() external {
+        for (uint32 i = 1; i < 6; i++) {
+            string[] memory cmds = new string[](3);
+            cmds[0] = "diff-test";
+            cmds[1] = "plonk-batch-verify";
+            cmds[2] = vm.toString(i);
+
+            bytes memory result = vm.ffi(cmds);
+            (
+                IPlonkVerifier.VerifyingKey[] memory verifyingKeys,
+                uint256[][] memory publicInputs,
+                IPlonkVerifier.PlonkProof[] memory proofs,
+                bytes[] memory extraTranscriptInitMsgs
+            ) = abi.decode(
+                result,
+                (IPlonkVerifier.VerifyingKey[], uint256[][], IPlonkVerifier.PlonkProof[], bytes[])
+            );
+
+            assert(V.batchVerify(verifyingKeys, publicInputs, proofs, extraTranscriptInitMsgs));
+        }
+    }
+
+    /// forge-config: default.fuzz.runs = 30
+    /// @dev Test when bad public inputs are supplied, the verification should fail
+    /// We know our `gen_circuit_for_test` in `diff_test.rs` has only 3 public inputs
+    function testFuzz_badPublicInputs_fails(uint256[3] calldata randPublicInput) external {
```

Comment:
Should we add a similar test with bad verifying keys?

Reply:
This is up for debate.
I used to have it, but if we consider the actual usage: VK is hardcoded contract that we ship our contract with, it's not a user-provided input, thus its integrity/correctness is not an issue.

OTOH, it's also fine to test the implementation of `batchVerify()` which should fail if there's an honest mistake in the process of hardcoding the VK contract. 🤔

Reply:
After some thoughts, I agree that we should add this bad path test, done in [b914f77](https://github.com/EspressoSystems/espresso-sequencer/pull/722/commits/b914f7761975e341ddd7f49eea9125f8f72efbd6)